distribution/push_v2.go

```diff
@@ -454,5 +515,108 @@ func layerAlreadyExists(ctx context.Context, metadata []metadata.V2Metadata, rep
+	candidates := []metadata.V2Metadata{}
+
+	for _, meta := range v2Metadata {
+		sourceRepo, err := reference.ParseNamed(meta.SourceRepository)
+		if err != nil {
+			continue
+		}
+		if repoInfo.Hostname() != sourceRepo.Hostname() {
+			continue
+		}
+		candidates = append(candidates, meta)
+	}
+
+	sortV2MetadataByLikenessAndAge(repoInfo, authConfigHash, candidates)
+
+	if max >= 0 && len(candidates) > max {
+		// select the youngest metadata
+		candidates = candidates[:max]
+	}
+
+	return candidates
+}
+
+// byLikenessAndAge is a sorting container for v2 metadata candidates for cross
+// repository mount. The candidate a is preferred over b:
+//
+//  1. if its AuthConfigHash matches the one used to authenticate to target repository
+//     and the hash of the "b" does not
+//  2. if a number of its repository path components exactly matching path components of target repository is higher
+//  3. if it was added later
+type byLikenessAndAge struct {
+	arr            []metadata.V2Metadata
+	authConfigHash string
+	pathComponents []string
+	// added[i] stands for an order of addition of item arr[i]
+	// the higher the number, the newer the entry
+	added []int
+}
+
+func (bla byLikenessAndAge) Less(i, j int) bool {
+	if bla.arr[i].AuthConfigHash == bla.authConfigHash && bla.arr[j].AuthConfigHash != bla.authConfigHash {
+		return true
+	}
+	if (bla.arr[i].AuthConfigHash == bla.authConfigHash) == (bla.arr[j].AuthConfigHash == bla.authConfigHash) {
+		aMatch := numOfMatchingPathComponents(bla.arr[i].SourceRepository, bla.pathComponents)
+		bMatch := numOfMatchingPathComponents(bla.arr[j].SourceRepository, bla.pathComponents)
+		if aMatch > bMatch {
+			return true
+		}
+		if aMatch == bMatch {
+			return bla.added[i] > bla.added[j]
+		}
+	}
+	return false
+}
+func (bla byLikenessAndAge) Swap(i, j int) {
+	bla.arr[i], bla.arr[j] = bla.arr[j], bla.arr[i]
+	bla.added[i], bla.added[j] = bla.added[j], bla.added[i]
+
+}
+func (bla byLikenessAndAge) Len() int { return len(bla.arr) }
+
+func sortV2MetadataByLikenessAndAge(repoInfo reference.Named, authConfigHash string, marr []metadata.V2Metadata) {
+	added := make([]int, len(marr))
+	for i := range marr {
+		added[i] = i
+	}
+	sort.Sort(byLikenessAndAge{
+		arr:            marr,
+		authConfigHash: authConfigHash,
+		added:          added,
+		pathComponents: strings.Split(repoInfo.Name(), "/"),
+	})
+}
+
+// numOfMatchingPathComponents returns a number of path components in "pth" that exactly match "matchComponents".
+func numOfMatchingPathComponents(pth string, matchComponents []string) int {
+	pthComponents := strings.Split(pth, "/")
+	i := 0
+	for ; i < len(pthComponents) && i < len(matchComponents); i++ {
+		if matchComponents[i] != pthComponents[i] {
+			return i
+		}
+	}
+	return i
+}
+
+func cancelLayerUpload(ctx context.Context, dgst digest.Digest, layerUpload distribution.BlobWriter) {
+	if layerUpload != nil {
+		go func() {
```

Comment:
I just wanted to speed up cross-repo mounting. I don't have a strong argument for doing this concurrently. I just felt that this would be a good optimization.